src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -39,21 +41,52 @@ private partial class CompilationTracker
                 /// </summary>
                 public readonly bool DocumentsAreFinal;
 
+                /// <summary>
+                /// Whether the generated documents are frozen and generators should never be ran again, ever, even if a document
+                /// is later changed. This is used to ensure that when we produce a frozen solution for partial semantics,
+                /// further downstream forking of that solution won't rerun generators. This is because of two reasons:
+                /// <list type="number">
+                /// <item>Generally once we've produced a frozen solution with partial semantics, we now want speed rather
+                /// than accuracy; a generator running in a later path will still cause issues there.</item>
+                /// <item>The frozen solution with partial semantics makes no guarantee that other syntax trees exist or
+                /// whether we even have references -- it's pretty likely that running a generator might produce worse results
+                /// than what we originally had.</item>
+                /// </list>
+                /// </summary>
+                public readonly bool DocumentsAreFinalAndFrozen;
+
                 public CompilationTrackerGeneratorInfo(
                     TextDocumentStates<SourceGeneratedDocumentState> documents,
                     GeneratorDriver? driver,
-                    bool documentsAreFinal)
+                    bool documentsAreFinal,
+                    bool documentsAreFinalAndFrozen = false)
```

Comment:
Would you consider using a enum here excessive? It seems to me that having something like `enum GeneratedDocumentsState` with 3 members `NotFinal`, `Final` and `FinalAndFrozen` would make it clearer. i.e. you can only change the state in one direction.

Reply:
Not a bad idea, OK to defer just to limit churn here?